resample.go

```diff
@@ -47,98 +49,103 @@ func ResampleRatio(quality int, ratio float64, s Streamer) *Resampler {
 	if quality < 1 || 64 < quality {
 		panic(fmt.Errorf("resample: invalid quality: %d", quality))
 	}
-	if math.IsInf(ratio, 0) || math.IsNaN(ratio) {
+	if ratio <= 0 || math.IsInf(ratio, 0) || math.IsNaN(ratio) {
 		panic(fmt.Errorf("resample: invalid ratio: %f", ratio))
 	}
 	return &Resampler{
 		s:     s,
 		ratio: ratio,
-		first: true,
-		buf1:  make([][2]float64, 512),
-		buf2:  make([][2]float64, 512),
+		buf1:  make([][2]float64, resamplerSingleBufferSize),
+		buf2:  make([][2]float64, resamplerSingleBufferSize),
 		pts:   make([]point, quality*2),
-		off:   0,
-		pos:   0,
+		off:   -resamplerSingleBufferSize,
+		pos:   0.0,
+		end:   math.MaxInt,
 	}
 }
 
 // Resampler is a Streamer created by Resample and ResampleRatio functions. It allows dynamic
 // changing of the resampling ratio, which can be useful for dynamically changing the speed of
 // streaming.
 type Resampler struct {
-	s          Streamer     // the orignal streamer
+	s          Streamer     // the original streamer
 	ratio      float64      // old sample rate / new sample rate
-	first      bool         // true when Stream was not called before
 	buf1, buf2 [][2]float64 // buf1 contains previous buf2, new data goes into buf2, buf1 is because interpolation might require old samples
 	pts        []point      // pts is for points used for interpolation
 	off        int          // off is the position of the start of buf2 in the original data
-	pos        int          // pos is the current position in the resampled data
+	pos        float64      // pos is the current position in the resampled data
+	end        int          // end is the position after the last sample in the original data
 }
 
 // Stream streams the original audio resampled according to the current ratio.
 func (r *Resampler) Stream(samples [][2]float64) (n int, ok bool) {
-	// if it's the first time, we need to fill buf2 with initial data, buf1 remains zeroed
-	if r.first {
-		sn, _ := r.s.Stream(r.buf2)
-		r.buf2 = r.buf2[:sn]
-		r.first = false
-	}
-	// we start resampling, sample by sample
 	for len(samples) > 0 {
-	again:
-		for c := range samples[0] {
-			// calculate the current position in the original data
-			j := float64(r.pos) * r.ratio
+		// Calculate the current position in the original data.
+		wantPos := r.pos * r.ratio
 
-			// find quality*2 closest samples to j and translate them to points for interpolation
-			for pi := range r.pts {
-				// calculate the index of one of the closest samples
-				k := int(j) + pi - len(r.pts)/2 + 1
+		// Determine the quality*2 closest sample positions for the interpolation.
+		// The window has length len(r.pts) and is centered around wantPos.
+		windowStart := int(wantPos) - (len(r.pts)-1)/2 // (inclusive)
```

Comment:
Good catch! But I prefer having the casts because it conveys on the same line as it it used that the value is rounded down, which is important to the equations. Performance-wise I don't think it will be noticeable in a benchmark.

Reply:
Yeah I'm good with that... In the grand scheme of things a couple casts isn't going to sink the ship :D And it's likely that the compiler would magic it away anyway

Reply:
I hope so, but I'm not going to disassemble it to check. :D